modules/client/src/internal/utils.ts

```diff
@@ -824,3 +863,801 @@ export function toPluginPermissionOperationType(
+  // length should be 5
+  if (
+    receivedPattern.length !== 5 ||
+    !compareArrays(receivedPattern, PLUGIN_UPDATE_WITH_ROOT_ACTION_PATTERN)
+  ) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Returns true if the actions are valid for a plugin update proposal without root permission
+ *
+ * @export
+ * @param {ProposalActionTypes[]} actions
+ * @return {*}  {boolean}
+ */
+export function isPluginUpdateAction(actions: ProposalActionTypes[]): boolean {
+  // get the first 3 action
+  // if the first is a dao upgrade skip it
+  let receivedPattern;
+  if (
+    actions[0] === ProposalActionTypes.UPGRADE_TO ||
+    actions[0] === ProposalActionTypes.UPGRADE_TO_AND_CALL
+  ) {
+    receivedPattern = actions.slice(1, 4);
+  } else {
+    receivedPattern = actions.slice(0, 3);
+  }
+  // check if it matches the expected pattern
+  // length should be 3
+  if (
+    receivedPattern.length !== 3 ||
+    !compareArrays(receivedPattern, PLUGIN_UPDATE_ACTION_PATTERN)
+  ) {
+    return false;
+  }
+  return true;
+}
+/**
+ * Returns true if the actions are valid for a plugin update proposal
+ *
+ * @export
+ * @param {ProposalActionTypes[]} actions
+ * @return {*}  {boolean}
+ */
+export function isDaoUpdateAction(actions: ProposalActionTypes[]): boolean {
+  // UpgradeTo or UpgradeToAndCall should be the first action
+  return actions[0] === ProposalActionTypes.UPGRADE_TO ||
+    actions[0] === ProposalActionTypes.UPGRADE_TO_AND_CALL;
+}
+
+export function validateUpdateDaoProposalActions(
+  actions: DaoAction[],
+  daoAddress: string,
+  implementationAddress: string,
+  currentDaoVersion: [number, number, number],
+): DaoUpdateProposalValidity {
+  const classifiedActions = classifyProposalActions(actions);
+  const causes: DaoUpdateProposalInvalidityCause[] = [];
+  // check if the actions are valid
+  // if they are not valid return add
+  // invalid actions to the causes and return
+  if (isDaoUpdateAction(classifiedActions)) {
+    // if they are valid, the upgrade action must
+    // be the first one
+    const upgradeActionType = classifiedActions[0];
+    const upgradeAction = actions[0];
+    // if the to address is not the dao address
+    // add the cause to the causes array
+    if (upgradeAction.to !== daoAddress) {
+      causes.push(
+        DaoUpdateProposalInvalidityCause.INVALID_TO_ADDRESS,
+      );
+    }
+    // if the value is different from 0
+    // add the cause to the causes array
+    if (upgradeAction.value.toString() !== "0") {
+      causes.push(
+        DaoUpdateProposalInvalidityCause.INVALID_VALUE,
+      );
+    }
+    // if the upgrade action is upgradeTo
+    if (upgradeActionType === ProposalActionTypes.UPGRADE_TO) {
+      // decode the upgradeTo action
+      const decodedImplementationAddress = decodeUpgradeToAction(
+        actions[0].data,
+      );
+      // check that the implementation address is the same
+      if (implementationAddress !== decodedImplementationAddress) {
```

Comment:
```suggestion
      if (expectedImplementationAddress !== decodedImplementationAddress) {
```
Where will the `expectedImplementationAddress` come from? From the active contracts by the App or SDK?

Reply:
From the dao factory address that the sdk have, if a specifica version is giveen it'll use that version to read the address in the LIVE_CONTRACTS